cpp/src/parquet/chunker_internal_test.cc

```diff
@@ -0,0 +1,1397 @@
+  ASSERT_OK_AND_ASSIGN(auto deleted, ConcatAndCombine({part1_, part2_, part3_, edit2_}));
+  ASSERT_FALSE(base->Equals(*deleted));
+  ASSERT_EQ(deleted->num_rows(), base->num_rows() - edit1_->num_rows());
+
+  ASSERT_OK_AND_ASSIGN(
+      auto base_info,
+      WriteAndGetParquetInfo(base, kMinChunkSize, kMaxChunkSize, kEnableDictionary,
+                             ParquetDataPageVersion::V1, kRowGroupLength));
+  ASSERT_OK_AND_ASSIGN(
+      auto deleted_info,
+      WriteAndGetParquetInfo(deleted, kMinChunkSize, kMaxChunkSize, kEnableDictionary,
+                             ParquetDataPageVersion::V1, kRowGroupLength));
+
+  ASSERT_EQ(base_info.size(), 7);
+  ASSERT_EQ(deleted_info.size(), 7);
+
+  ASSERT_EQ(base_info.at(0).page_lengths, deleted_info.at(0).page_lengths);
+  ASSERT_EQ(base_info.at(1).page_lengths, deleted_info.at(1).page_lengths);
+  for (size_t i = 2; i < deleted_info.size() - 1; i++) {
+    AssertPageLengthDifferences(base_info.at(i), deleted_info.at(i),
+                                /*exact_number_of_equal_diffs=*/0,
+                                /*exact_number_of_larger_diffs=*/1,
+                                /*exact_number_of_smaller_diffs=*/1, edit1_->num_rows());
+  }
+  AssertPageLengthDifferences(base_info.back(), deleted_info.back(),
+                              /*exact_number_of_equal_diffs=*/0,
+                              /*exact_number_of_larger_diffs=*/0,
+                              /*exact_number_of_smaller_diffs=*/1, edit1_->num_rows());
+}
+
+TEST_F(TestCDCMultipleRowGroups, UpdateOnce) {
+  auto constexpr kRowGroupLength = 128 * 1024;
+  auto constexpr kEnableDictionary = false;
+  auto constexpr kMinChunkSize = 0 * 1024;
+  auto constexpr kMaxChunkSize = 128 * 1024;
+
+  ASSERT_OK_AND_ASSIGN(auto base,
+                       ConcatAndCombine({part1_, edit1_, part2_, part3_, edit2_}));
+  ASSERT_OK_AND_ASSIGN(auto updated,
+                       ConcatAndCombine({part1_, edit3_, part2_, part3_, edit2_}));
+  ASSERT_FALSE(base->Equals(*updated));
+
+  ASSERT_OK_AND_ASSIGN(
+      auto base_info,
+      WriteAndGetParquetInfo(base, kMinChunkSize, kMaxChunkSize, kEnableDictionary,
+                             ParquetDataPageVersion::V1, kRowGroupLength));
+  ASSERT_OK_AND_ASSIGN(
+      auto updated_info,
+      WriteAndGetParquetInfo(updated, kMinChunkSize, kMaxChunkSize, kEnableDictionary,
+                             ParquetDataPageVersion::V1, kRowGroupLength));
+
+  ASSERT_EQ(base_info.size(), 7);
+  ASSERT_EQ(updated_info.size(), 7);
+
+  ASSERT_EQ(base_info.at(0).page_lengths, updated_info.at(0).page_lengths);
+  ASSERT_EQ(base_info.at(1).page_lengths, updated_info.at(1).page_lengths);
+  AssertPageLengthDifferences(base_info.at(2), updated_info.at(2),
+                              /*max_number_of_equal_diffs=*/1);
+  for (size_t i = 2; i < updated_info.size(); i++) {
+    ASSERT_EQ(base_info.at(i).page_lengths, updated_info.at(i).page_lengths);
+  }
```

Comment:
Why call `AssertPageLengthDifferences` if we use `ASSERT_EQ` just after for `i = 2`?

Reply:
`AssertPageLengthDifferences` checks the page differences within a row group. We expect a single diff in the third row group updating values but not shifting any of the pages because the number of the values are the same (otherwise it would be an update + insertion/deletion). All the other row groups should be identical.

Added a comment.